src/wire/report.rs

```diff
@@ -0,0 +1,412 @@
+        le_bytes::write_u64(&mut data[field::SEQ_NUM], value);
+    }
+
+    /// Set the `n_clocks` field
+    #[inline]
+    pub fn set_n_clocks(&mut self, value: u16) {
+        let data = self.buffer.as_mut();
+        le_bytes::write_u16(&mut data[field::N_CLOCKS], value);
+    }
+
+    /// Set the `n_log_entries` field
+    #[inline]
+    pub fn set_n_log_entries(&mut self, value: u32) {
+        let data = self.buffer.as_mut();
+        le_bytes::write_u32(&mut data[field::N_LOG_ENTRIES], value);
+    }
+
+    /// Return a mutable pointer to the payload
+    #[inline]
+    pub fn payload_mut(&mut self) -> &mut [u8] {
+        &mut self.buffer.as_mut()[field::PAYLOAD.start..]
+    }
+}
+
+impl<T: AsRef<[u8]>> AsRef<[u8]> for WireReport<T> {
+    fn as_ref(&self) -> &[u8] {
+        self.buffer.as_ref()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[rustfmt::skip]
+    static MSG_BYTES: [u8; 54] = [
+        // fingerprint
+        0x54, 0x50, 0x52, 0x4D,
+        // probe_id: 1
+        0x01, 0x00, 0x00, 0x00,
+        // clock: 2
+        0x02, 0x00, 0x00, 0x00,
+        // seq_id: 8
+        0x08, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+        // n_clocks: 2
+        0x02, 0x00,
+        // n_log_entries: 3
+        0x03, 0x00, 0x00, 0x00,
+        // payload
+        0x01, 0x00, 0x00, 0x00,
+        0x02, 0x00, 0x00, 0x00,
+        0x03, 0x00, 0x00, 0x00,
+        0x04, 0x00, 0x00, 0x00,
+        0x05, 0x00, 0x00, 0x00,
+        0x06, 0x00, 0x00, 0x00,
+        0x07, 0x00, 0x00, 0x00,
+    ];
+
+    #[rustfmt::skip]
+    static PAYLOAD_BYTES: [u8; 28] = [
+        // logical_clock[0]
+        0x01, 0x00, 0x00, 0x00,
+        0x02, 0x00, 0x00, 0x00,
+        // logial_clock[1]
+        0x03, 0x00, 0x00, 0x00,
+        0x04, 0x00, 0x00, 0x00,
+        // log_entry[0]
+        0x05, 0x00, 0x00, 0x00,
+        // log_entry[1]
+        0x06, 0x00, 0x00, 0x00,
+        // log_entry[2]
+        0x07, 0x00, 0x00, 0x00,
+    ];
+
+    #[test]
+    fn header_len() {
+        assert_eq!(WireReport::<&[u8]>::header_len(), 26);
+        let n_clocks = 12;
+        let n_log_items = 14;
+        assert_eq!(
+            WireReport::<&[u8]>::buffer_len(n_clocks, n_log_items),
+            26 + (12 * mem::size_of::<LogicalClock>()) + (14 * mem::size_of::<LogEntry>())
+        );
+    }
+
+    #[test]
+    fn construct() {
+        let mut bytes = [0xFF; 54];
+        let mut r = WireReport::new_unchecked(&mut bytes[..]);
+        assert_eq!(r.check_len(), Ok(()));
+        r.set_fingerprint();
+        r.set_probe_id(ProbeId::new(1).unwrap());
+        r.set_clock(2);
+        r.set_seq_num(8);
+        r.set_n_clocks(2);
+        r.set_n_log_entries(3);
+        r.payload_mut().copy_from_slice(&PAYLOAD_BYTES[..]);
+        assert_eq!(r.check_fingerprint(), Ok(()));
+        assert_eq!(r.check_payload_len(), Ok(()));
+        assert_eq!(&r.into_inner()[..], &MSG_BYTES[..]);
+    }
+
+    #[test]
+    fn deconstruct() {
```

Comment:
Could we have an API-oriented struct-independent round-tripping test? One where we wrap a byte array with WireReport, set the field values, then hand over the byte array to a fresh WireReport instance and confirm that the newly-accessible values match what was originally written? This would be a backstop against accidental statefulness of the WireReport instance and how that might affect interpretability.